src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataCollectionSerializer.cs

```diff
@@ -90,6 +91,12 @@ public ODataCollectionSerializer(ODataSerializerProvider serializerProvider)
                 {
                     collectionStart.NextPageLink = writeContext.InternalRequest.Context.NextLink;
                 }
+                else if (writeContext.InternalRequest.Context.QueryOptions != null)
+                {
+                    // Collection serializer is called only for collection of primitive values - A null object will be supplied since it is a non-entity value
+                    SkipTokenHandler skipTokenHandler = SkipTokenQueryOption.GetSkipTokenHandler(writeContext.QueryOptions.Context);
+                    collectionStart.NextPageLink = skipTokenHandler.GenerateNextPageLink(writeContext.InternalRequest.RequestUri, writeContext.InternalRequest.Context.PageSize, null, writeContext);
```

Comment:
>NextPageLink [](start = 36, length = 12)

shall we do like "ResourceSet" to write the NextpageLink when call "WriteEnd()" ?

Reply:
Seems like the ResourceSet approach is a work around. We can discuss this today.

---
In reply to: [281402176](https://github.com/OData/WebApi/pull/1724#discussion_r281402176) [](ancestors = 281402176)